NIFI-4809 - Implement a SiteToSiteMetricsReportingTask

pvillard31 opened the pull request:
To avoid some code duplication, I moved few utilitary classes into the reporting-utils package. And I also added two metrics (available cores and load average) as we have in System Diagnostic but that we don't send with the current Ambari reporting task.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

pvillard31 commented:
I'll need to update this PR if #2431 is merged first.

mattyb149 commented:
Reusing the Ambari format certainly makes this one easier to implement, and as you said the user can convert later with a record processor, but I'm thinking it might be best to be able to specify the RecordWriter in the reporting task. That way the conversion, filtering, etc. can be done while the records are already in object format (i.e. parsed), which would save the step of re-parsing and re-writing. Using a JsonRecordSetWriter with "Inherit Record Schema" would result in the same behavior as you propose here, or you could provide a different format and/or filter out the fields, etc. by using a configured RecordSetWriter.

@markap14 suggested to encode the "input" schema as a file in src/main/resources and read it in when the class/instance is created, using the available util method to convert it to a RecordSchema. To be honest I would like to have the S2S Provenance reporting task do this as well, but that's a bit more invasive since it already exists. Since the one in this PR is new, would be nice to have it be the exemplar for creating reporting services in the future. What do you think?

pvillard31 commented:
That's a great idea! I'll have a look.

mattyb149 commented on nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteMetricsReportingTask.java:
```diff
@@ -0,0 +1,151 @@
+    static final String TIMESTAMP_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+
+    static final PropertyDescriptor APPLICATION_ID = new PropertyDescriptor.Builder()
+            .name("Application ID")
```
Please set this(and the other properties) to a machine-friendly name, you can use the current value for .displayName()

mattyb149 commented on nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteMetricsReportingTask.java:
```diff
@@ -0,0 +1,151 @@
+
+    static final PropertyDescriptor APPLICATION_ID = new PropertyDescriptor.Builder()
+            .name("Application ID")
+            .description("The Application ID to be included in the metrics sent to Ambari")
```
The phrase "sent to Ambari" shows up a couple times, but should be replaced/removed

mattyb149 commented on nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteMetricsReportingTask.java:
```diff
@@ -0,0 +1,151 @@
+
+                final Map<String, String> attributes = new HashMap<>();
+                final String transactionId = UUID.randomUUID().toString();
+                attributes.put("reporting.task.transaction.id", transactionId);
```
Now that #2431 is in, we should add those attributes here too :)

pvillard31 commented:
Thanks for your comments @mattyb149 - I just pushed a commit that should address everything.

Regarding the record approach you suggested. Even though I really like the idea, I'm not sure to see how to define a valid avro schema for the specification used by the Ambari collector API (https://cwiki.apache.org/confluence/display/AMBARI/Metrics+Collector+API+Specification):

````json
{
  "metrics": [
    {
      "metricname": "AMBARI_METRICS.SmokeTest.FakeMetric",
      "appid": "amssmoketestfake",
      "hostname": "ambari20-5.c.pramod-thangali.internal",
      "timestamp": 1432075898000,
      "starttime": 1432075898000,
      "metrics": {
        "1432075898000": 0.963781711428,
        "1432075899000": 1432075898000
      }
    }
  ]
}
````

How would we manage the 'metrics' part where field names are timestamps?

mattyb149 commented:
Ugh that's true, I'm not a fan of dynamic keys. Since the schema would be generated by the reporting task, then we could create a schema for the example above, but then each flow file would have its own schema, and even worse, each flow file would have only one metric, so at that point it's not really conducive to record processing. As an alternative we could convert the output (before or after JSON conversion) to an altered spec with a consistent schema definition. 
For the Ambari reporting task, since Ambari is expecting this format, then fine; if we keep the same spec in this new reporting task for consistency, then I'd hope to see a template on the Wiki using the new reporting task with a JoltTransformJSON processor to do the aforementioned transformation, along with including an AvroSchemaRegistry that contains the schema definition for the files coming out of the JoltTransformJSON processor. This would allow us to keep a consistent (standard) defined format (albeit non-schema-friendly), but offer a well-known solution to prepare the data for record-aware processors. Thoughts?

mattyb149 commented:
I like all those ideas!

pvillard31 commented:
@mattyb149 - this is not the final version, I still have unit tests and "additional details" doc to add, but wanted to give you an update in case you want to have a look and if you already have feedbacks.

I've reworked the new reporting task as I suggested with two additional properties. I also reworked the way metrics are constructed to have the ambari format and the "record" format with the minimal code duplication. I also chose this implementation (using the JsonTreeReader) so that we can easily reuse the code for the other S2S reporting task. If it sounds OK to you, adding an optional Record Writer property to the other S2S reporting tasks should be really easy (I guess best is to take care of it into a follow-up JIRA).

pvillard31 commented:
Done @mattyb149, thanks!

mattyb149 commented on nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/AbstractSiteToSiteReportingTask.java:
```diff
@@ -140,8 +164,16 @@
+    static final PropertyDescriptor RECORD_WRITER = new PropertyDescriptor.Builder()
+            .name("record-writer")
+            .displayName("Record Writer")
+            .description("Specifies the Controller Service to use for writing out the records.")
```
The description here should mention that it is only used when the reporting task is configured to use a Record Writer. For the one you added, it has two modes, but future reporting tasks need not offer an alternative.

pvillard31 replied:
Left it as-is since I added the property Record Writer only in the task reporting task for now.

mattyb149 commented on nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteMetricsReportingTask.java:
```diff
@@ -0,0 +1,212 @@
+    static final PropertyDescriptor FORMAT = new PropertyDescriptor.Builder()
+            .name("s2s-metrics-format")
+            .displayName("Output format")
+            .description("The output format that will be used for the metrics")
```
Perhaps add to the description that if Record Format is selected, a Record Writer must be provided. Also for consistency, the display name should have both words capitalized.

pvillard31 replied:
done

mattyb149 commented on nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteMetricsReportingTask.java:
```diff
@@ -0,0 +1,212 @@
+        final List<ValidationResult> problems = new ArrayList<>(super.customValidate(validationContext));
+
+        final boolean isWriterSet = validationContext.getProperty(RECORD_WRITER).isSet();
+        if (validationContext.getProperty(FORMAT).getValue().equals(RECORD_FORMAT.getValue()) && !isWriterSet) {
```
Can we also check that a Record Writer is NOT set when Ambari Format is selected? This isn't a requirement, just a suggestion. Since it requires two operations to switch back and forth between formats, that's not an ideal user experience, so I'm fine with whichever way you'd like to do it :)

pvillard31 replied:
I added the check and added unit tests for the customValidate

mattyb149 commented on nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/resources/docs/org.apache.nifi.reporting.SiteToSiteMetricsReportingTask/additionalDetails.html:
```diff
@@ -0,0 +1,178 @@
+        <link rel="stylesheet" href="../../../../../css/component-usage.css" type="text/css" />
+    </head>
+
+    <body>
```
This is excellent documentation, thank you!

mattyb149 commented:
+1 LGTM, ran the unit tests and tried on a live NiFi instance with both Ambari and Record modes. The only thing I noticed (that should be its own Jira) is that the RecordWriter used by the reporting task does not report any referencing components.
Thanks for this great addition! Merging to master
